src/main/java/com/google/sps/servlets/MealServlet.java

```diff
@@ -177,26 +179,47 @@ private void getMealById(HttpServletRequest request, HttpServletResponse respons
     }
 
     private void returnIdOfSimilar(HttpServletRequest request, HttpServletResponse response) throws IOException {
-        // Fuction redirect to random page
-        // TODO(grenlayk): implement suggestions algorithm here for Product Alpha
+        // Fuction redirect to random page of the same type
         Long mealId = Long.parseLong(getParameter(request, "id", "0"));
-
-        Query query = new Query("Meal");
+        Meal meal = null;
+        Filter idFilter = new FilterPredicate("id", FilterOperator.EQUAL, mealId);
+        Query query = new Query("Meal").setFilter(idFilter);
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-        PreparedQuery results = datastore.prepare(query);
-        ArrayList<Long> idList = new ArrayList<>();
-        for (Entity entity : results.asIterable()) {
-            Long id = (Long)entity.getProperty("id");
-            if (id != null && !id.equals(mealId)) {
-                idList.add(id);
-            }
+        List<Meal> meals = getDataFromDatastore(datastore.prepare(query)); 
+        if (meals.size() > 1) {
+            response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+            return;
+        } else if (meals.size() == 1) {
+            meal = (Meal)meals.get(0);
+        } else {
+            response.sendError(HttpServletResponse.SC_NOT_FOUND);
+            return;
         }
+
+        Filter anotherIdFilter = new FilterPredicate("id", FilterOperator.NOT_EQUAL, meal.getId());
+        Filter typeFilter = new FilterPredicate("type", FilterOperator.EQUAL, meal.getType());
+        CompositeFilter sameTypeFilter =
+            CompositeFilterOperator.and(anotherIdFilter, typeFilter);
+        Query typeQuery = new Query("Meal").setFilter(sameTypeFilter);
+        Query idQuery = new Query("Meal").setFilter(anotherIdFilter);
+        List<Meal> sameTypeIdList = getDataFromDatastore(datastore.prepare(typeQuery));
+        List<Meal> idList = getDataFromDatastore(datastore.prepare(idQuery));
+
         Random rand = new Random(); 
-        int index = rand.nextInt(idList.size());
-        Long randomId = idList.get(index);
+        int index = 0;
+        Meal randomMeal = null;
+        if (sameTypeIdList.size() > 0) {
+            index = rand.nextInt(sameTypeIdList.size());
+            randomMeal = (Meal)sameTypeIdList.get(index);
```

Comment:
I guess you can write a function that would just pick a random item from a list.